src/gwt/src/org/rstudio/studio/client/workbench/UserInterfaceHighlighter.java

```diff
@@ -70,19 +91,109 @@ public Element getHighlightElement()
       {
          return highlightedElement_;
       }
-      
+
+      public void clearHandler()
+      {
+         if (handler_ != null)
+            handler_.removeHandler();
+      }
+
+      public void executeCallback()
+      {
+         // This method must be called by a single HighlightPair, but because there can be multiple
+         // pairs per query, we need to check if the callback has already executed before proceeding.
+         if(!highlighter_.getCallbackProcessed(index_))
```

Comment:
Can you elaborate a bit on what happens if you have multiple UI elements highlighted, and the callback associated with one is executed? (What happens to the other highlighted UI elements?)

Main reason I ask: normally, one will highlight multiple UI elements either because:

1. There are multiple pieces of UI associated with the same action or command (e.g. multiple buttons for saving a document). In this case, one might want to clear all highlighters after the associated command has been executed (e.g. via click or via shortcut);

2. Some of the highlighted UI might act like "bread crumbs"; e.g. highlighting a button + menu item leading to the "final" item to be clicked or used. In this case, you normally wouldn't want to clear any highlight until the user has clicked on the highlight associated with the clickable button.

I'm not sure whether this orchestration should happen on the RStudio side, or if it should be a requirement of the code handling the callback.

Reply:
Each callback is associated with a single query that may have multiple elements - if any of these elements are selected then the handlers for all the elements per that query are cleared. My understanding is to make case #2 possible the user would need to send a list of queries, so they would only pass the callback with the final query. 

In case #1 if they are sending multiple queries with the same callback, then the callback will only be executed and removed from the element the user selects. Once the user clears the highlights, the other listeners will be removed as well. 

This implementation makes intuitive sense to me, but there may be something I'm missing.

Reply:
Got it; I think that makes sense. It sounds like both cases are possible under the existing framework; it just affects how the user submits their highlight request through `rstudioapi`. Does that sound correct?

In each case, it seems like the ultimate goal of the callback would be to call `rstudioapi::highlightUi(list())` to clear any existing highlight after the appropriate user interaction had taken place.

Reply:
Yes, that's correct! There is a chance if multiple elements have the same callback and the highlights don't get cleared that the callback code would run more than once, but I think that's the reasonable expected outcome.